harness/determined/common/experimental/resource_pool.py

```diff
@@ -0,0 +1,113 @@
+from typing import List, Optional
+
+from determined.common import api
+from determined.common.api import bindings
+from determined.common.experimental import workspace
+
+
+class ResourcePool:
+    """A class representing a resource pool object.
+
+    Attributes:
+        name: (str) The name of the resource pool.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        name: Optional[str] = None,
+    ):
+        """Create a resource pool object.
+
+        Arguments:
+            session: The session to use for API calls.
+            workspace_name: (Optional) Name of the resource pool.
+        """
+        self.name = name
+        self._session = session
+
+    @classmethod
+    def _from_bindings(
+        cls, resource_pool_bindings: bindings.v1ResourcePool, session: api.Session
+    ) -> "ResourcePool":
+        resource_pool = cls(session, name=resource_pool_bindings.name)
+        return resource_pool
+
+    def bind(
+        self,
+        workspace_names: List[str],
+    ) -> None:
+        """
+        Bind a resource pool to workspaces.
+
+        Arguments:
+            workspace_names (list(str)): The names of the workspaces to be bound.
+        """
+        req = bindings.v1BindRPToWorkspaceRequest(
+            resourcePoolName=self.name,
+            workspaceNames=workspace_names,
+        )
+
+        bindings.post_BindRPToWorkspace(self._session, body=req, resourcePoolName=self.name)
+
+    def unbind(
+        self,
+        workspace_names: List[str],
+    ) -> None:
+        """
+        Unbind a resource pool from workspaces.
+
+        Arguments:
+            workspace_names (list(str)): The names of the workspaces to be unbound.
+        """
+        req = bindings.v1UnbindRPFromWorkspaceRequest(
+            resourcePoolName=self.name,
+            workspaceIds=workspace_names,
+        )
+
+        bindings.delete_UnbindRPFromWorkspace(self._session, body=req, resourcePoolName=self.name)
+
+    def workspaces(self) -> List[str]:
+        """
+        List workspaces bound to a resource pool.
+
+        Returns:
+            (List(str)) The names of workspaces bound to the resource pool.
+        """
+
+        def get_with_offset(offset: int) -> bindings.v1ListWorkspacesBoundToRPResponse:
+            return bindings.get_ListWorkspacesBoundToRP(
+                session=self._session,
+                offset=offset,
+                resourcePoolName=self.name,
+            )
+
+        resps = api.read_paginated(get_with_offset)
+        workspace_names = [
+            workspace.Workspace(session=self._session, workspace_id=w).workspace_name
+            for r in resps
+            if r.workspaceIds is not None
```

Comment:
This conditional is surprising to me. When does it happen?

Reply:
I think this is because resps can be more than one response due to pagination, so we have to check each response to make sure workspaceIds is populated?

Reply:
I put the check there for two reasons.
1. The Optional type you mentioned above.
2. Being conservative, I'm not sure what it returns if there is no workspace bound to the RP. An empty list or None? I should have probably played with it and read the docs.

Reply:
@wes-turner Good catch. The optional is due a misconfig in the proto file.
```
message ListWorkspacesBoundToRPResponse {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: { required: [ "workspace_id", "limit" ] }
  };
  // List of workspace IDs.
  repeated int32 workspace_ids = 1;
  // Pagination information of the full dataset.
  Pagination pagination = 2;
}
```
```
message GetExperimentsResponse {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: { required: [ "experiments", "pagination" ] }
  };
  // The list of returned experiments.
  repeated determined.experiment.v1.Experiment experiments = 1;
  // Pagination information of the full dataset.
  Pagination pagination = 2;
}
```